c10/util/complex.h

```diff
@@ -0,0 +1,477 @@
+  explicit constexpr complex_common(const std::complex<U> &other): complex_common(other.real(), other.imag()) {}
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  explicit C10_HOST_DEVICE complex_common(const thrust::complex<U> &other): complex_common(other.real(), other.imag()) {}
+#endif
+
+  constexpr complex<T> &operator =(T re) {
+    storage[0] = re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  constexpr complex<T> &operator +=(T re) {
+    storage[0] += re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  constexpr complex<T> &operator -=(T re) {
+    storage[0] -= re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  constexpr complex<T> &operator *=(T re) {
+    storage[0] *= re;
+    storage[1] *= re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  constexpr complex<T> &operator /=(T re) {
+    storage[0] /= re;
+    storage[1] /= re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator =(const complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator +=(const complex<U> &rhs) {
+    storage[0] += rhs.real();
+    storage[1] += rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator -=(const complex<U> &rhs) {
+    storage[0] -= rhs.real();
+    storage[1] -= rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator *=(const complex<U> &rhs) {
+    // (a + bi) * (c + di) = (a*c - b*d) + (a * d + b * c) i
+    T a = storage[0];
+    T b = storage[1];
+    U c = rhs.real();
+    U d = rhs.imag();
+    storage[0] = a * c - b * d;
+    storage[1] = a * d + b * c;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator /=(const complex<U> &rhs) {
+    // (a + bi) / (c + di) = (ac + bd)/(c^2 + d^2) + (bc - ad)/(c^2 + d^2) i
+    T a = storage[0];
+    T b = storage[1];
+    U c = rhs.real();
+    U d = rhs.imag();
+    auto denominator = c * c + d * d;
+    storage[0] = (a * c + b * d) / denominator;
+    storage[1] = (b * c - a * d) / denominator;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator =(const std::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  C10_HOST_DEVICE complex<T> &operator =(const thrust::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+#endif
+
+  template<typename U>
+  explicit constexpr operator std::complex<U>() const {
+    return std::complex<U>(std::complex<T>(real(), imag()));
```

Comment:
lol

Reply:
This provides the behavior that, if `std::complex<T>` can not be casted to `std::complex<U>`, the `c10::complex<T>` won't either.

Reply:
In the scenario, ```std::complex<T>``` can not be casted to ```std::complex<U>```, it would fail with that error message instead of "c10::complex<T> can't be casted to ... ". is that desirable?

Reply:
I think it will say that no conversion from T to U for both case:

```C++
#include <complex>

struct T {};

int main() {
    static_cast<std::complex<T>>(std::complex<double>());
}
```

```
In file included from <source>:1:

/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/complex: In instantiation of 'constexpr std::complex<_Tp>::complex(const std::complex<_Up>&) [with _Up = double; _Tp = T]':

<source>:6:57:   required from here

/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/complex:141:43: error: no matching function for call to 'T::T(double)'

  141 |  : _M_real(__z.real()), _M_imag(__z.imag()) { }

      |                                           ^

<source>:3:8: note: candidate: 'constexpr T::T()'

    3 | struct T {};

      |        ^

<source>:3:8: note:   candidate expects 0 arguments, 1 provided

<source>:3:8: note: candidate: 'constexpr T::T(const T&)'

<source>:3:8: note:   no known conversion for argument 1 from 'double' to 'const T&'

<source>:3:8: note: candidate: 'constexpr T::T(T&&)'

<source>:3:8: note:   no known conversion for argument 1 from 'double' to 'T&&'

In file included from <source>:1:

/opt/compiler-explorer/gcc-9.3.0/include/c++/9.3.0/complex:141:43: error: no matching function for call to 'T::T(double)'

  141 |  : _M_real(__z.real()), _M_imag(__z.imag()) { }

      |                                           ^

<source>:3:8: note: candidate: 'constexpr T::T()'

    3 | struct T {};

      |        ^

<source>:3:8: note:   candidate expects 0 arguments, 1 provided

<source>:3:8: note: candidate: 'constexpr T::T(const T&)'

<source>:3:8: note:   no known conversion for argument 1 from 'double' to 'const T&'

<source>:3:8: note: candidate: 'constexpr T::T(T&&)'

<source>:3:8: note:   no known conversion for argument 1 from 'double' to 'T&&'

Compiler returned: 1
```